QoD Test definitions

jlurien opened the pull request:
#### What type of PR is this?

* enhancement/feature
* tests

#### What this PR does / why we need it:

This PR includes te first proposal for Test definitions which is aligned with the new Commonalities guidelines. Hving at least a basic test plan is a requirement to be part of a meta-release.


#### Which issue(s) this PR fixes:

Fixes #245 

#### Special notes for reviewers:

The proposal is quite extensive and covers more than just the minimum "basic" test plan, but this level of work is required to detect parts of the spec that are ambiguous or where implementation guidelines do not still define an explicit behaviour.

The README summarizes the list of comments in the Test definitions that require a review. Some of them may require as well a clarification in the API spec or in Commonalities guidelines. All of them can be considered not necessary for a basic test plan so they may be addressed in a subsequent version.

There is a feature file per API operation. All feature files are under the same folder /Test_definitions, but we may consider creating subfolders per API.

#### Changelog input

```
* Added test definitions for the 3 APIs
```

hdamker commented on code/Test_definitions/quality-on-demand-createSession.feature:
```diff
@@ -0,0 +1,374 @@
+Feature: CAMARA Quality On Demand API, v0.10.0-rc.1 - Operation createSession
```
Are the changes between `v0.10.0-rc.1` and `v0.11.0-rc.2` already considered?

jlurien replied:
Yes, it's a typo. Versions were aligned to r1.1. And they should be adapted to r1.2 as part of the PR, as with the API specs

RandyLevensalor commented on code/Test_definitions/qos-profiles-qosProfilesDevice.feature:
```diff
@@ -0,0 +1,224 @@
+
+    Background: Common qosProfilesDevice setup
+        Given an environment at "apiRoot"
+        And the resource "qos-profiles/v0.11rc1/qos-profiles"
```
```suggestion
        And the resource "qos-profiles/v0.11rc1/retrieve-qos-profiles"
```
Include the latest changes to the api.

jlurien replied:
Done

jlurien commented:
@hdamker I have updated the versions and paths to the right ones for r1.2 (without references to RC), and also updated the operationId name to retrieveQoSProfiles, as commented by @RandyLevensalor 

If no more comments are received, the PR would be ready to be merged before the deadline.

hdamker commented on code/Test_definitions/quality-on-demand-extendQosSessionDuration.feature:
```diff
@@ -0,0 +1,194 @@
+        And the response header "x-correlator" has same value as the request header "x-correlator"
+        And the response header "Content-Type" is "application/json"
+        And the response property "$.status" is 403
+        And the response property "$.code" is "PERMISSION_DENIED"
```
I agree that the generic "PERMISSION_DENIED" is appropriate here, but the API definitions lists also the more specific "INVALID_TOKEN_CONTEXT" for this scenario.

jlurien replied:
INVALID_TOKEN_CONTEXT was intended for device mismatch between the access_token and the request content. In this case the provided access token may be valid but the accessed resource is not accessible for the client, so maybe PERMISSION_DENIED is more common. I can allow both and add a note to the readme and test plan

hdamker commented on code/Test_definitions/quality-on-demand-deleteSession.feature:
```diff
@@ -0,0 +1,126 @@
+        And the response header "x-correlator" has same value as the request header "x-correlator"
+        And the response header "Content-Type" is "application/json"
+        And the response property "$.status" is 403
+        And the response property "$.code" is "PERMISSION_DENIED"
```
same comment as in extendQosSessionDuration

jlurien replied:
I have applied the same in all the affected scenarios (6 in total)

hdamker commented on code/Test_definitions/qos-profiles-retrieveQoSProfiles.feature:
```diff
@@ -0,0 +1,224 @@
+    # * A device object identifying a device commercialized by the implementation for which the service is not applicable, if any
+
+
+    # References to OAS spec schemas refer to schemas specifies in qos-profiles.yaml, version 0.11.0-rc.1
```
Here `-rc.1` is wrong as the renaming to `retrieve-qos-profiles` happened after -rc.1 :-)

 ```suggestion
    # References to OAS spec schemas refer to schemas specifies in qos-profiles.yaml, version 0.11.0
```

hdamker commented:
> If no more comments are received, the PR would be ready to be merged before the deadline.

@jlurien I have looked over the quality-on-demand feature files and they are looking really good to me - as far as it is possible to see without implementing the test and with limited knowledge of the language. I'm ready to approve, just some minor formal comments above.

jlurien commented:
@hdamker I hope I have correctly addressed your feedback, please take another look

hdamker reviewed (dismissed):
LGTM - thanks a lot @jlurien for this huge contribution and the swift corrections. 

If there are issues during implementation or clarifications of the open points we have still the option to release patches of the test definitions.

hdamker commented:
@jlurien Selected results from the manual linting which I asked @rartych to run, which you might want to check:

```
qod-provisioning-createProvisioning.feature:
116    Scenario Outline does not have any Examples                                                       no-scenario-outlines-without-examples

qod-provisioning-deleteProvisioning.feature:
77     Step "When the asynchronous deletion process is completed" should not appear after step using keyword then    keywords-in-logical-order

qod-provisioning-retrieveProvisioningByDevice.feature:
78    Scenario Outline does not have any Examples    no-scenario-outlines-without-examples

quality-on-demand-extendQosSessionDuration.feature:
74     Scenario name is already used in: quality-on-demand-extendQosSessionDuration.feature:63    no-dupe-scenario-names
 97     Scenario name is already used in: quality-on-demand-extendQosSessionDuration.feature:86    no-dupe-scenario-names

hdamker commented on code/Test_definitions/quality-on-demand-deleteSession.feature:
```diff
@@ -0,0 +1,127 @@
+    @quality_on_demand_deleteSession_403.1_session_token_mismatch
+    Scenario: QoS session not created by the API client given in the access token
+        # To test this, a token have to be obtained for a different client
+        Given the header "Authorization" is set to a valid access token emitted to a client which did not created the QoS session        And the header "Authorization" is set to a valid access token emitted for a different device
```
Another one from linting, was raised as "Step name is too long", but actually the line feed was missing (if I got it right):

```suggestion
        Given the header "Authorization" is set to a valid access token emitted to a client which did not created the QoS session        
        And the header "Authorization" is set to a valid access token emitted for a different device
```

jlurien replied:
That second Authorization should not be there, that's why is too long. I've removed it

jlurien commented:
> @jlurien Selected results from the manual linting which I asked @rartych to run, which you might want to check:
> 
> ```
> qod-provisioning-createProvisioning.feature:
> 116    Scenario Outline does not have any Examples                                                       no-scenario-outlines-without-examples
> 
> qod-provisioning-deleteProvisioning.feature:
> 77     Step "When the asynchronous deletion process is completed" should not appear after step using keyword then    keywords-in-logical-order
> 
> qod-provisioning-retrieveProvisioningByDevice.feature:
> 78    Scenario Outline does not have any Examples    no-scenario-outlines-without-examples
> 
> quality-on-demand-extendQosSessionDuration.feature:
> 74     Scenario name is already used in: quality-on-demand-extendQosSessionDuration.feature:63    no-dupe-scenario-names
>  97     Scenario name is already used in: quality-on-demand-extendQosSessionDuration.feature:86    no-dupe-scenario-names
> ```

I've corrected the relevant ones:

- Scenario Outline without examples, totally
- Step too long was really due to a previous step appended which had to be removed
- Repeated scenario names
- Multiple empty lines, not really wrong but same as trailing spaces
- The 2 When in an scenario is not really an error, but a common best practice. I have rephrase it a bit because for that specific scenario makes sense (first you delete and then when the async deletion is effective you received an event)

I have not changed:

- Filenames were really OK according to the guidelines (For APIs with several operations that split the test plan in one feature file per operation, recommendation is to add the operationId to the api-name as file name: quality-on-demand-createSession.feature.) But operationId are lowerCamelCase
- Scenario too long. Good practice but sporadically some scenarios may need some more steps

hdamker commented on code/Test_definitions/quality-on-demand-deleteSession.feature:
```diff
@@ -0,0 +1,126 @@
+    # * The sessionId of an existing session with status "AVAILABLE", and with provided values for "sink" and "sinkCredential".
+    # * The sessionId of an existing session with status "UNAVAILABLE", and with provided values for "sink" and "sinkCredential".
+    #
+    # References to OAS spec schemas refer to schemas specifies in quality-on-demand.yaml, version 0.11.0-rc.1
```
Still one more :-) 

```suggestion
    # References to OAS spec schemas refer to schemas specifies in quality-on-demand.yaml, version 0.11.0
```

jlurien replied:
ups

hdamker commented:
Good to go here - keeping the milestone date! Again a big thank you @jlurien
